src/engine/blocks.js

```diff
@@ -80,6 +87,34 @@ Blocks.prototype.getOpcode = function (id) {
     return this._blocks[id].opcode;
 };
 
+/**
+ * Get all fields and their values for a block.
+ * @param {?string} id ID of block to query.
+ * @return {!Object} All fields and their values.
+ */
+Blocks.prototype.getFields = function (id) {
+    if (typeof this._blocks[id] === 'undefined') return null;
+    return this._blocks[id].fields;
+};
+
+/**
+ * Get all non-substack inputs for a block.
+ * @param {?string} id ID of block to query.
+ * @return {!Object} All non-substack inputs and their associated blocks.
+ */
+Blocks.prototype.getInputs = function (id) {
+    if (typeof this._blocks[id] === 'undefined') return null;
+    var inputs = {};
+    for (var input in this._blocks[id].inputs) {
+        // Ignore blocks prefixed with substack prefix.
+        if (input.substring(0, Blocks.SUBSTACK_INPUT_PREFIX.length)
+            != Blocks.SUBSTACK_INPUT_PREFIX) {
```

Comment:
I think the Boolean flag would make sense, since like you suggested it avoids the need for additional complexity in `moveBlock` and maybe other code in the future.

Are there other parts of the code that could benefit from such a flag? On the one hand I'm in favor of explicit data (like an `isStatement` or `inputType` field or whatever) over implicit data (special knowledge about the prefix of some input names), but on the other hand I'm concerned that I'm making a big deal about one substring compare.